### ISSUE TITLE
Failed to join thread, despite catch_unwind!: Any

### ISSUE THREAD

**Issue by Savir** [target/lorri#375]
**Describe the bug**

**To Reproduce**

1. started lorry as a user daemon (`systemctl --user restart lorri`)
2. logged into a directory with a `default.nix` config file
3. check the logs `journalctl --user -fu lorri    `

**Expected behavior**

**Metadata**

```
Apr 13 13:58:51 hector-tests systemd[1579]: lorri.service: Main process exited, code=exited, status=101/n/a
Apr 13 13:58:51 hector-tests systemd[1579]: lorri.service: Failed with result 'exit-code'.
Apr 13 13:58:51 hector-tests systemd[1579]: lorri.service: Service hold-off time over, scheduling restart.
Apr 13 13:58:51 hector-tests systemd[1579]: lorri.service: Scheduled restart job, restart counter is at 1.
Apr 13 13:58:51 hector-tests systemd[1579]: Stopped Lorri Daemon.
Apr 13 13:58:51 hector-tests systemd[1579]: Started Lorri Daemon.
Apr 13 13:58:51 hector-tests lorri.sh[10958]: Apr 13 13:58:51.268 INFO ready
Apr 13 14:02:05 hector-tests lorri.sh[10958]: Apr 13 14:02:05.409 INFO build status, message: BuildEvent(Started { nix_file: Shell("/home/hector/Documents/tests/Projects/test/shell.nix"), reason: ProjectAdded(Shell("/home/hector/Documents/tests/Projects/test/shell.nix")) })                                                                                                       
Apr 13 14:03:44 hector-tests lorri.sh[10958]: Well, this is embarrassing.
Apr 13 14:03:44 hector-tests lorri.sh[10958]: lorri had a problem and crashed. To help us diagnose the problem you can send us a crash report.
Apr 13 14:03:44 hector-tests lorri.sh[10958]: We have generated a report file at "/tmp/report-99e37834-4263-41c8-85d5-180afe58cc98.toml". Submit an issue or email with the subject of "lorri Crash Report" and include the report as an attachment.
Apr 13 14:03:44 hector-tests lorri.sh[10958]: - Homepage: https://github.com/target/lorri
Apr 13 14:03:44 hector-tests lorri.sh[10958]: - Authors: Graham Christensen <graham.christensen@target.com>
Apr 13 14:03:44 hector-tests lorri.sh[10958]: We take privacy seriously, and do not perform any automated error collection. In order to improve the software, we rely on people to submit reports.
Apr 13 14:03:44 hector-tests lorri.sh[10958]: Thank you kindly!

```

```
$ cat /tmp/report-24b4aa0e-f2af-4d5c-8b6b-dcc5d56eb97f.toml                                                                                                                    
name = 'lorri'
operating_system = 'unix:Ubuntu'
crate_version = '1.0.0'
explanation = '''
Panic occurred in file 'src/libcore/result.rs' at line 1188
'''
cause = 'Failed to join thread, despite catch_unwind!: Any'
method = 'Panic'
backtrace = '''

   0: 0x55ba3828b0f1 - lorri::daemon::Daemon::serve::h668a22fd0bedf273
   1: 0x55ba38325873 - lorri::ops::daemon::main::ha222b79ed8a6d05d
   2: 0x55ba382387c0 - lorri::main::hf56b324794c28b07
   3: 0x55ba38248673 - std::rt::lang_start::{{closure}}::h434019b262ad67f2
   4: 0x55ba384c52a3 - std::panicking::try::do_call::h098c78edce79182c
   5: 0x55ba384db3fa - __rust_maybe_catch_panic
   6: 0x55ba384cde11 - std::rt::lang_start_internal::h8b246e89f17ae792
   7: 0x55ba3823a1e2 - main
   8: 0x7fb31b599d8b - __libc_start_main
   9: 0x55ba382301ea - _start
  10:        0x0 - <unknown>'''
```

```console
$ lorri info --shell-file ./default.nix                                                                                                                                                                                    
lorri version: 1.0
gc roots do not exist, shell_gc_root: "/home/hector/.cache/lorri/gc_roots/16ab78779e8a73512f8d98b3100fb330/gc_root/shell_gc_root"
```

```console
$ uname -a
Linux hector-tests 5.3.0-46-generic #38~18.04.1-Ubuntu SMP Tue Mar 31 04:17:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Additional context**

**Comment by Profpatsch** [comment]
Hm, I’m afraid I can’t see any helpful information in the stack trace apart from the fact that some thread didn’t want to finish.

Can you share the `shell.nix` that you tried to evaluate, and maybe filter it down to a minimal example that still crashes?

You can run the daemon outside of the service, simply run `lorri daemon` in a terminal. Maybe with `-vv` to get more output.

**Comment by aveltras** [comment]
Hi, I encounter the same issue when trying to use lorri with an obelisk project https://github.com/obsidiansystems/obelisk.
Minimum reproducible shell.nix as follows:
`let

  nixpkgsRev = "fce7562cf46727fdaf801b232116bc9ce0512049";
  pkgs = import (githubTarball "NixOS" "nixpkgs-channels" nixpkgsRev) {};
  
  githubTarball = owner: repo: rev:
    builtins.fetchTarball { url = "https://github.com/${owner}/${repo}/archive/${rev}.tar.gz"; };

  obelisk = import (githubTarball "obsidiansystems" "obelisk" "v0.8.0.0") {};
  
in pkgs.mkShell {
  buildInputs = with pkgs; [
    (haskell.lib.justStaticExecutables obelisk.command)
  ];
}`

**Comment by curiousleo** [comment]
@aveltras thanks for the `shell.nix`, it's always good to have repro cases. However, the stacktrace in your message is different from the one in the ticket description. I believe yours is a different issue. Would you mind creating a separate ticket? Please include your operating system in the description.

"permission denied" could point to a known host of issues with Nix on macOS. I was unable to reproduce @aveltras' issue on NixOS 19.09.

**Comment by Profpatsch** [comment]
Okay, I got the same issue now.

The stack trace points to the thread pool join

https://github.com/target/lorri/blob/33cb219d8bea17dd400789eec220f0a1bdea313c/src/daemon.rs#L156

which is a small thread pool implementation in the lorri codebase:

https://github.com/target/lorri/blob/33cb219d8bea17dd400789eec220f0a1bdea313c/src/thread.rs#L83-L105

The code is a bit older, and we exchanged the `std::sync::mpsc` channels with `crossbeam_channel` in the meantime, which could be a reason for it failing now cc @curiousleo.

However, the implementation might have been buggy all along cc @grahamc, we should take a closer look at whether it does what it should. I’m not entirely sure about the semantics of how (and when) panics are intended to be propagated from the threads in the pool. The `loop` of the join method I linked above would never receive a message from a panicked thread as far as I can see.

Are we sure that `drop` is called on the `DeathCertificate` if a thread panics?

https://github.com/target/lorri/blob/33cb219d8bea17dd400789eec220f0a1bdea313c/src/thread.rs#L19-L25

According to to `join` documentation it panics if the thread to be `join`ed is panicing, so this looks like the intended behavior to me, only that the error message we throw with expect is strange (what’s `catch_unwind` mean?). It feels like the crash just means that one of the threads in the pool panicked, but doesn’t propagate the panic message.

**Comment by Profpatsch** [comment]
There is a small no-dependency crate called `scoped_threadpool` which seems to be doing pretty much the same thing, maybe it makes sense to replace our implementation with it: http://kimundi.github.io/scoped-threadpool-rs/src/scoped_threadpool/lib.rs.html#217-243

**Comment by curiousleo** [comment]
I'm experimenting with `scoped_threadpool` here: https://github.com/target/lorri/pull/433

**Comment by curiousleo** [comment]
Fixed in #435.
